pkg/agent/util/nettest/packetconn.go

```diff
@@ -47,37 +47,36 @@ func NewPacketConn(localAddr net.Addr, inCh chan *Packet, outCh chan *Packet) *P
 }
 
 func (pc *PacketConn) ReadFrom(p []byte) (int, net.Addr, error) {
+	// Check if connection is closed once before the select statement. Otherwise we may end up
+	// reading a packet even if the connection was already closed before calling this
+	// function. It is still possible for the connection to be closed between this check and the
+	// select, but it doesn't matter in this case because that would mean the 2 function calls
+	// (Close and ReadFrom) are concurrent.
```

Comment:
To confirm if I understand it correctly, it doesn't matter because the test scenarios don't make concurrent calls?

Reply:
No that's not what I meant. It doesn't matter because if the calls are concurrent, there are not ordered. If the calls (`Close` / `ReadFrom`) are concurrent, then it is fine for `ReadFrom` to return a packet.

Prior to this change (the second commit), it was possible for a single goroutine to execute `Close`, followed by `ReadFrom`, and the `ReadFrom` could potentially return a packet. This ie because the second select will execute a case chosen at random if both are "ready". This was not a correct implementation of the `PacketConn` interface, but the new version is a correct implementation. There is not assumption as to how the tests are implemented, and this `PacketConn` implementation should be flexible enough to work for "arbitrary" tests.

Reply:
Got it, I missed the randomness of the select statement, thanks for the explanation.